### PR TITLE
OpenGL 3.0 based lens distortion shaders

### PR DESCRIPTION
**Pull request by BastiaanOlij**
Because OpenGL 1/2 and OpenGL 3 don't always mix on all platforms I had to make slightly altered versions of the shaders that run on OpenGL 3 while making OpenHMD work with Godot.

Figured, might as well submit them here as they may help out others who are going down the OpenGL 3 route.

### CODE REVIEW

File: src/shaders.c [hunk 1]
```diff
@@ -55,3 +55,80 @@ const char * const distortion_frag =
     "//Black edges off the texture\n"
     "gl_FragColor = ((tc_g.x < 0.0) || (tc_g.x > 1.0) || (tc_g.y < 0.0) || (tc_g.y > 1.0)) ? vec4(0.0, 0.0, 0.0, 1.0) : vec4(red, green, blue, 1.0);\n"
 "}";
+
+// OpenGL 3.0 compatible shaders, because not all platforms allow you to use OpenGL 1/2 style stuff anymore when enabling OpenGL 3...
+// You need an empty vao for this, no need to load up vertex buffer. This is assuming you limit left/right with glViewport. 
+// glBindVertexArray(vao);
+// glDrawArrays(GL_TRIANGLES, 0, 3 * 2);
+// glBindVertexArray(0);
```
Comment (Wallbraker) [inline src/shaders.c]: This comment should probably go into one of the doccomments in `include/openhmd.h`.

File: include/openhmd.h [hunk 2]
```diff
@@ -61,6 +61,8 @@ typedef enum {
 typedef enum {
 	OHMD_GLSL_DISTORTION_VERT_SRC = 0,
 	OHMD_GLSL_DISTORTION_FRAG_SRC = 1,
+	OHMD_GLSL_DISTORTION_VERT_OGL3_SRC = 2,
+	OHMD_GLSL_DISTORTION_FRAG_OGL3_SRC = 3,
```
Comment (Wallbraker) [inline include/openhmd.h]: Should probably be called `OHMD_GLSL_330_DISTORTION...` so people know exactly which `#version <number>` header we start the shaders with.

File: src/shaders.c (hunk 1 above)
Comment (BastiaanOlij) [inline src/shaders.c]: I'll take it out for now, it really needs an example application. Problem is that OpenGL 3 setup is a bit more involving then the simple OpenGL 1 example that exist right now

**Comment by jsarrett** [comment]
Hardcoded verts in the vertex shader seems like a bad idea.  For example that now makes and assumption about canvas orientation of the HMD, which is up to the windows system.  I intentionally didn't want to know anything about the viewport transform the user set up for their render pipeline.  I don't know what is customary among GL3+developers, but I'd prefer if the verts and the MVP transform were passed into the shader.

**Comment by jsarrett** [comment]
Also, is ist possible with preprocessor directives to keep it as one shader source, instead of two?  i.e. inlining the differences, so that the basic arithmetic isn't duplicated?  maybe not?  I'm not sure, but it sure would be good if we don't have to repeat that code.  Maybe it could be done by pulling out the model into a function and calling it from different  main programs where are behind #if __GLSL_VERSION__ guards?

**Comment by BastiaanOlij** [comment]
The reason the hardcoded verts are more common in OpenGL 3 code for small applications such as this is because there is far more setup required. glMatrixMode, glBegin, glEnd, glTexCoord2D, glVertex3d they have all been deprecated and on Mac and OGL ES for instance, aren't even available anymore the moment you use an OpenGL 3 or later context.

Adding an MVP in is easy. You simple add a uniform called MVP and maybe that is enough to give the flexibility you require, so I'll add that in tonight.

Loading up the list of vertices with their coordinates adds an awful lot of code that the person using OpenHMD suddenly has to write while the whole idea is to already solve all that and offer finished building blocks. You have to define a structure for holding the verts, create a vertex buffer, configure the vertex arrays, all for 4 vertices?

The problem with a compiler directive is that you would need to compile different versions for different platforms, though I can see the elegance too because you will never use the two versions in unison. 

What I have done in my own engine (not in Godot that I'm working on atm, but something I've done for myself), is write a simple shader parser so I can add directives into the shader code itself. That would be a far more elegant solution.
It actually makes sense to me to add the code to compile the shaders into OpenHMD and maybe even a helper method to run the shader and it would be pretty easy to get a simple parser running.

**Comment by BastiaanOlij** [comment]
Ok, the shader has been changed to use an MVP matrix and VBO for the vertices. I didn't implement any preprocessor or anything to lessen the duplication of having the two versions of the shaders. It's getting late.

Actually using the shaders have a look at the code I wrote for using OpenHMD for Godot here:
https://github.com/BastiaanOlij/godot/blob/openhmd_3.0/modules/openhmd/openhmd_shader.cpp
That calss encapsulates all the logic that compiles, configures and uses the OpenGL 3 shader.

**Comment by noname22** [comment]
Are we satisfied with this now @Wallbraker ?

**Comment by Wallbraker** [comment]
Thanks for the changes and patience!

**Pull request merged**